plugins/mana/plugin.go

```diff
@@ -371,6 +395,91 @@ func GetPendingMana(value float64, n time.Duration) float64 {
 	return value * (1 - math.Pow(math.E, -mana.Decay*(n.Seconds())))
 }
 
+// GetPastConsensusManaVector builds a consensus base mana vector in the past.
+func GetPastConsensusManaVector(t time.Time) (*mana.ConsensusBaseManaVector, error) {
+	cbmvPast := &mana.ConsensusBaseManaVector{}
+	cachedObj := consensusBaseManaPastVectorMetadataStorage.Get([]byte(mana.ConsensusBaseManaPastVectorMetadataStorageKey))
+	cachedMetadata := &mana.CachedConsensusBasePastManaVectorMetadata{CachedObject: cachedObj}
+	defer cachedMetadata.Release()
+
+	if cachedMetadata.Exists() {
+		metadata := cachedMetadata.Unwrap()
+		if t.After(metadata.Timestamp) {
+			consensusBaseManaPastVectorStorage.ForEach(func(key []byte, cachedObject objectstorage.CachedObject) bool {
+				cachedPbm := &mana.CachedPersistableBaseMana{CachedObject: cachedObject}
+				defer cachedPbm.Release()
+				p := cachedPbm.Unwrap()
+				err := cbmvPast.FromPersistable(p)
+				if err != nil {
+					log.Errorf("error while restoring %s mana vector from storage: %w", mana.ConsensusMana.String(), err)
+					cbmvPast = &mana.ConsensusBaseManaVector{}
+					return false
+				}
+				return true
+			})
+		}
+	}
+
+	var eventLogs []mana.Event
+	var readErr error
+	consensusEventsLogStorage.ForEach(func(key []byte, cachedObject objectstorage.CachedObject) bool {
+		cachedPe := &mana.CachedPersistableEvent{CachedObject: cachedObject}
+		pe := cachedPe.Unwrap()
+		defer cachedPe.Release()
+		if pe.Time.After(t) {
+			return false
+		}
+
+		// already consumed in stored base mana vector.
+		if cachedMetadata.Exists() && cbmvPast.Size() > 0 {
+			metadata := cachedMetadata.Unwrap()
+			if pe.Time.Before(metadata.Timestamp) {
+				return true
+			}
+		}
+
+		ev, err := mana.FromPersistableEvent(pe)
+		if err != nil {
+			readErr = err
+			return false
+		}
+		eventLogs = append(eventLogs, ev)
+		return true
+	})
+	if readErr != nil {
+		return nil, readErr
+	}
+	index, err := cbmvPast.BuildPastBaseVector(eventLogs, t)
+	if err != nil {
+		return nil, err
+	}
+
+	// store the most recent bmv
+	if err = consensusBaseManaPastVectorStorage.Prune(); err != nil {
+		return nil, err
+	}
+	for _, p := range cbmvPast.ToPersistables() {
+		consensusBaseManaPastVectorStorage.Store(p).Release()
+	}
```

Comment:
let's make this optional with a parameter `storePastReference` or smthing. Imagine that with this tool now we have the power to check if nodes had the exact same consensus mana perception for example 1 minute ago. But the main usecase for this being the calculation of ACMS after the end of the epochs.
With the following scenario:
 1. `GetPastConsensusManaVector()` gets called for `t`. (for anything else than epoch calc.)
 2. `consensusBaseManaPastVectorStorage` contains the bmv at `t`.
 3. `GetPastConsensusManaVector()` gets called for `t-10`. (for epoch calculation)
 4. Since `t-10` is before `t`, we skip building the initial past vector from `consensusBaseManaPastVectorStorage` entries, and have to consume ALL logged events.
 
So we should only store the latest bmv "checkpoint" when we are fairly sure that we will not need stuff from before (epoch calcualted). Ofc you could still query older timestamps, but that would be a fairly expensive operation compared to consuming only events since the checkpoint.

Another solution would be to have multiple checkpoints stored, let's say every X mins or X hours, then the amount of events you need to process would be determined by TPS.

Reply:
As discussed, we store only the latest bmv every X events.